File: 02_第一阶段-java加强-kotlindsl-函数式编程-gradle-git-高阶函数-lambda表达式/day05/code/src/com/itcast/ay05/接口回调/Mother.java
package com.itcast.ay05.接口回调;

/**
 * ClassName:Mother
 * Description:妈妈
 */
class Mother {
    public static void main(String[] args) {
        //有超市
        SuperMarket superMarket = new SuperMarket();

        //买酱油  买的时候没有任何延迟
        Son son = new Son();
//        son.feed(Soy("海天"));
        superMarket.buySoy(son);

//        System.out.println("买到了"+soy.name+"牌的酱油");
        //做菜

        System.out.println("做甜点");
    }
    //其实我们就是用feed这个方法函数 但是对于java来说 函数必须存在对象中
    //所以我们虽然只需要这个函数 ,但是必须要通过对象包裹来调用
    //kotlin函数可以单独存在 就可以直接通过函数来进行回调
    //第一步:定义能力
    //取物品的能力
    interface FeedBack {
        //取酱油行为
        void feed(Soy soy);
    }

    //儿子  取酱油
    static class Son implements FeedBack {
        //第四步:在具有能力的对象方法下进行接收
        @Override
        public void feed(Soy soy) {
            //取回了酱油
            System.out.println("获取到酱油了" + soy.name);
            System.out.println("妈妈开始做菜");
        }
    }
}

File: 02_第一阶段-java加强-kotlindsl-函数式编程-gradle-git-高阶函数-lambda表达式/day05/code/src/com/itcast/ay05/函数回调/Mother.kt
package com.itcast.ay05.函数回调


/**
 * ClassName:Mother
 * Description:mother
 */
fun main(args: Array<String>) {
    //创建超市对象
    val superMarket = SuperMarket()
    //买酱油
    //lambda表达式
//    superMarket.buySoy{
//        println("取回了${it.name}酱油")
//        println("妈妈开始做菜")
//    }

    superMarket.buySoy(null)

//    println("买到${soy.name}牌酱油")

    println("做甜点")
}